### PR TITLE
Implement certificate lookup, closes #24

### LINKED ISSUE #24: Add custom TLS certificate chain lookup
We can tell from experience that using prebuilt libcurl binaries can lead to problems when used outside of the distribution they were built for. Most commonly, it cannot find the system CA chain as the paths vary quite drastically across platforms.

In our officially released AppImages, we embed Alpine Linux's static build of libcurl. We should investigate whether we need a custom, cross-platform certificate chain lookup to make our tool work reliably across a defined (minimum, tested) set of platforms.

Note: we could even add some automated smoke testing on a variety of distributions with Docker on GitHub actions.

As an alternative, we could ship a CA chain as a fallback. In the past, we always considered that to be a UX quality degradation and strongly preferred using the system chains, though.

@probonopd please link to the related point in your desktop platform issues list.

### PR DESCRIPTION
**Pull request by probonopd**
Tested on openSUSE.

* Use `SSL_CERT_FILE` and/or `SSL_CERT_DIR` environment variables if set, otherwise check for common locations
* Do not check certificates if `APPIMAGETOOL_FETCH_RUNTIME_INSECURE` is set

References:
* https://curl.haxx.se/docs/sslcerts.html
* https://gitlab.com/probono/platformissues#certificates
* https://go.dev/src/crypto/x509/root_linux.go

### CODE REVIEW

**Comment by probonopd** [comment]
https://github.com/AppImage/appimagetool/suites/14153608220/artifacts/792967402

**Comment by probonopd** [comment]
Still doesnt work properly:

```
Using architecture x86_64
/home/linux/squashfs-root should be packaged as appimagetool-x86_64.AppImage
Generating squashfs...
Using certificate file /etc/ssl/ca-bundle.pem
Using certificate directory /etc/ssl/certs
Downloading runtime file from https://github.com/AppImage/type2-runtime/releases/download/continuous/runtime-x86_64
HEAD request to https://github.com/AppImage/type2-runtime/releases/download/continuous/runtime-x86_64 failed: error setting certificate file: /etc/ssl/certs/ca-certificates.crt
Failed to download runtime file
```

**Comment by probonopd** [comment]
Tested successfully on openSUSE Leap 15.5

File: README.md [hunk 1]
```diff
@@ -42,6 +42,9 @@ Some of the parameters above can alternatively be specified as environment varia
 - `APPIMAGETOOL_FORCE_SIGN`: By default, if signing fails, appimagetool just logs a warning but will not abort immediately. If this environment variable is set, appimagetool exits with a non-zero return code.
 - `APPIMAGETOOL_SIGN_PASSPHRASE`: If the `--sign-key` is encrypted and requires a passphrase to be used for signing (and, for some reason, GnuPG cannot be used interactively, e.g., in a CI environment), this environment variable can be used to safely pass the key.
 - `VERSION`: This value will be inserted by appimagetool into the root desktop file and (if the destination parameter is not provided by the user) in the output filename.
+- `APPIMAGETOOL_FETCH_RUNTIME_INSECURE`: If set, CA certificates will not be used when downloading the runtime. 
+- `SSL_CERT_FILE`: If set, CA certificates will be looked up in this file when downloading the runtime.
```
Comment (TheAssassin) [inline README.md:46]: Should be `CA_*` (or maybe even `APPIMAGETOOL_CA_*`).

File: README.md [hunk 2]
```diff
@@ -42,6 +42,9 @@ Some of the parameters above can alternatively be specified as environment varia
 - `APPIMAGETOOL_FORCE_SIGN`: By default, if signing fails, appimagetool just logs a warning but will not abort immediately. If this environment variable is set, appimagetool exits with a non-zero return code.
 - `APPIMAGETOOL_SIGN_PASSPHRASE`: If the `--sign-key` is encrypted and requires a passphrase to be used for signing (and, for some reason, GnuPG cannot be used interactively, e.g., in a CI environment), this environment variable can be used to safely pass the key.
 - `VERSION`: This value will be inserted by appimagetool into the root desktop file and (if the destination parameter is not provided by the user) in the output filename.
+- `APPIMAGETOOL_FETCH_RUNTIME_INSECURE`: If set, CA certificates will not be used when downloading the runtime.
```
Comment (TheAssassin) [inline README.md:45]: We should really emphasize the security issues.

... maybe we should just *not* mention this option at all? Or even remove it completely?

File: src/appimagetool.c [hunk 3]
```diff
@@ -882,6 +882,9 @@ main (int argc, char *argv[])
         if (verbose)
             printf("Size of the embedded runtime: %d bytes\n", size);
         
+        if (verbose)
```
Comment (TheAssassin) [inline src/appimagetool.c:885]: This change is completely unrelated. You should either move your commit to a new branch and send a separate PR or clean up your Git history so we can merge it as a part of this PR.

... also, I'm not sure whether it's redundant to the other messages.

File: .gitignore [hunk 4]
```diff
@@ -3,3 +3,4 @@ cmake-build-*/
 *.*swp*
 *.bak
 *.AppImage*
+squashfs-root/
```
Comment (TheAssassin) [inline .gitignore:6]: Trailing line break missing.

File: src/appimagetool_fetch_runtime.c [hunk 5]
```diff
@@ -13,6 +15,69 @@
...
+
+            for (size_t i = 0; i < sizeof(certDirectories) / sizeof(certDirectories[0]); ++i) {
+                if (access(certDirectories[i], F_OK) != -1) {
+                    if (curl_easy_setopt(handle, CURLOPT_CAPATH, certDirectories[i]) == CURLE_OK) {
+                        fprintf(stderr, "Using certificate directory %s\n", certDirectories[i]);
+                        break;
+                    }
+                }
+            }
+        }
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:79]: All of this code should be in a separate function. This is pure bloat to this one.

File: src/appimagetool_fetch_runtime.c [hunk 6]
```diff
@@ -13,6 +15,69 @@
 bool fetch_runtime(char* arch, size_t* size, char** buffer, bool verbose) {
     // not the cleanest approach to globally init curl here, but this method shouldn't be called more than once anyway
     curl_global_init(CURL_GLOBAL_ALL);
+    CURL* handle = curl_easy_init();
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:18]: I'm not sure whether it's okay to reuse such a handle.

File: src/appimagetool_fetch_runtime.c [hunk 7]
```diff
@@ -13,6 +15,69 @@
...
+
+    if (handle == NULL) {
+        fprintf(stderr, "Failed to initialize libcurl\n");
+        curl_global_cleanup();
+        return false;
+    }
+
+    if (getenv("APPIMAGETOOL_FETCH_RUNTIME_INSECURE") != NULL) {
+        curl_easy_setopt(handle, CURLOPT_SSL_VERIFYPEER, 0L);
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:27]: There should be a *very*, **very**, ***very*** bold warning log message.

File: src/appimagetool_fetch_runtime.c [hunk 8]
```diff
@@ -13,6 +15,69 @@
...
+            }
+        } else {
+            curl_easy_setopt(handle, CURLOPT_CAINFO, NULL);
+            const char* certFiles[] = {
+            "/etc/ssl/certs/ca-certificates.crt",                // Debian/Ubuntu/Gentoo etc.
+            "/etc/pki/tls/certs/ca-bundle.crt",                  // Fedora/RHEL 6
+            "/etc/ssl/ca-bundle.pem",                            // OpenSUSE
+            "/etc/pki/tls/cacert.pem",                           // OpenELEC
+            "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem", // CentOS/RHEL 7
+            "/etc/ssl/cert.pem",                                 // Alpine Linux
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:47]: Formatting is off. These lines should be indented.

File: src/appimagetool_fetch_runtime.c [hunk 9]
```diff
@@ -13,6 +15,69 @@
...
+            };
+            for (size_t i = 0; i < sizeof(certFiles) / sizeof(certFiles[0]); ++i) {
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:49]: Might be easier to just use a `NULL`-terminated list above. This works of course, but it might be more readable. You don't need any numeric index then, but can just `++` your pointer until you encounter a `NULL` value. We do this a lot in this code.

File: src/appimagetool_fetch_runtime.c [hunk 10]
```diff
@@ -13,6 +15,69 @@
...
+                }
+            }
+        }
+
+        if (getenv("SSL_CERT_DIR") != NULL) {
+            if (curl_easy_setopt(handle, CURLOPT_CAPATH, getenv("SSL_CERT_DIR")) == CURLE_OK) {
+                fprintf(stderr, "Using certificate directory %s\n", getenv("SSL_CERT_DIR"));
+            }
+        } else {
+            curl_easy_setopt(handle, CURLOPT_CAPATH, NULL);
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:64]: Why? You do not need to `NULL` it explicitly. Code like this needs comments that explain the rationale.

File: src/appimagetool_fetch_runtime.c [hunk 11]
```diff
@@ -13,6 +15,69 @@
...
+            const char* certDirectories[] = {
+                "/etc/ssl/certs",               // SLES10/SLES11, https://golang.org/issue/12139
+                "/etc/pki/tls/certs",           // Fedora/RHEL
+                "/system/etc/security/cacerts", // Android
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:68]: I'm not saying we absolutely need to remove it, but do we really need *Android* support for this tool?

Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:68]: Please avoid the use of in-line comments. Especially ones aligned like these are just annoying to work with.

File: src/appimagetool_fetch_runtime.c [hunk 12]
```diff
@@ -86,9 +150,6 @@ bool fetch_runtime(char* arch, size_t* size, char** buffer, bool verbose) {
         fprintf(stderr, "HEAD request to %s failed: %s\n", url, curl_error_buf);
     }
 
-    curl_easy_cleanup(handle);
-    handle = NULL;
```
Comment (TheAssassin) [inline src/appimagetool_fetch_runtime.c:90]: The code below assumes in its error handling that the handle has been cleaned up. If you _really_ want to reuse it, you need to continuously call `curl_easy_cleanup` wherever the function returns.

... this is why we need C++ sooner than later. I consider adding some before the 1.0.0 release. Handling object lifetime correctly is much easier there due to RAII.

**Review by TheAssassin — CHANGES_REQUESTED** [review]
This is not an in-depth code review. Let's call it a workflow review. When all of the points are discussed, I'll have another look.

**Comment by TheAssassin** [comment]
See #32 for an alternative implementation. This alternative version does not support the environment variables proposed in this PR. In my opinion, it is a viable workaround to require from users to manually download the runtime if the TLS validation fails. And if it's a bug, we want to motivate them to create an issue this way.
